Add logging sampling documentation

evgenyfedorov2 opened the pull request:
## Summary

Fixes #45661 


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/extensions/log-sampling.md](https://github.com/dotnet/docs/blob/0064b6c216b17ee24c7b19e14d1fe3336649b68a/docs/core/extensions/log-sampling.md) | [Log sampling in .NET](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/log-sampling?branch=pr-en-us-45948) |
| [docs/fundamentals/toc.yml](https://github.com/dotnet/docs/blob/0064b6c216b17ee24c7b19e14d1fe3336649b68a/docs/fundamentals/toc.yml) | [docs/fundamentals/toc](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/toc?branch=pr-en-us-45948) |
| [docs/navigate/tools-diagnostics/toc.yml](https://github.com/dotnet/docs/blob/0064b6c216b17ee24c7b19e14d1fe3336649b68a/docs/navigate/tools-diagnostics/toc.yml) | [docs/navigate/tools-diagnostics/toc](https://review.learn.microsoft.com/en-us/dotnet/navigate/tools-diagnostics/toc?branch=pr-en-us-45948) |


<!-- PREVIEW-TABLE-END -->

evgenyfedorov2 committed:
Add logging sampling docs

evgenyfedorov2 committed:
Update toc

gfoidl commented on docs/core/extensions/snippets/logging/logging-sampling/Program.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace LogSamplingCodeConfig;
+
+internal static class Program
+{
+    public static void Main()
```
Why not use top level statements?

---

The same on other places (I don't comment on them to reduce noise)

evgenyfedorov2 replied:
Done

gfoidl commented on docs/core/extensions/snippets/logging/logging-sampling/code-config/LogSamplingCodeConfig.csproj:
```diff
@@ -0,0 +1,21 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="appsettings.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
```
Isn't this the default from the SDK?
So this could be omitted for brevity.

---

The same on other places (I don't comment on them to reduce noise)

evgenyfedorov2 replied:
No, by default is not copied to the` bin/Debug` folder nor taken into account if located in the same folder as the `.cspro`j file. At least that's how it behaves on my machine :) 

Anyway, I removed the file from projects where it is not really required.

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/file-config/Program.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace LogSamplingFileConfig;
+
+internal static class Program
+{
+    public static void Main()
+    {
+        var hostBuilder = Host.CreateApplicationBuilder();
+        hostBuilder.Logging.AddSimpleConsole(options =>
+        {
+            options.SingleLine = true;
+            options.TimestampFormat = "hh:mm:ss";
+        });
+
+        // Add the Random probabilistic sampler to the logging pipeline.
+        hostBuilder.Logging.AddRandomProbabilisticSampler(hostBuilder.Configuration);
+
+        var app = hostBuilder.Build();
+        var loggerFactory = app.Services.GetRequiredService<ILoggerFactory>();
+        var logger = loggerFactory.CreateLogger("SamplingDemo");
+
+        while (true)
+        {
+            Log.EnteredWhileLoop(logger);
+            Log.NoisyMessage(logger);
+            Log.LeftWhileLoop(logger);
+
+            Thread.Sleep(100);
+        }
+    }
+}
```
```suggestion
var hostBuilder = Host.CreateApplicationBuilder();

hostBuilder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
hostBuilder.Logging.AddRandomProbabilisticSampler(hostBuilder.Configuration);

using var app = hostBuilder.Build();

var loggerFactory = app.Services.GetRequiredService<ILoggerFactory>();
var logger = loggerFactory.CreateLogger("SamplingDemo");

while (true)
{
    Log.EnteredWhileLoop(logger);
    Log.NoisyMessage(logger);
    Log.LeftWhileLoop(logger);

    await Task.Delay(100);
}
```

evgenyfedorov2 replied:
Done

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/trace-based/Log.cs:
```diff
@@ -0,0 +1,12 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
```suggestion
```

evgenyfedorov2 replied:
Done

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/trace-based/LogSamplingTraceBased.csproj:
```diff
@@ -0,0 +1,24 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Hosting" Version="9.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="9.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Telemetry" Version="9.4.0" />
```
```suggestion
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="9.0.4" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="9.0.4" />
    <PackageReference Include="Microsoft.Extensions.Telemetry" Version="9.4.0" />
```

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/trace-based/LogSamplingTraceBased.csproj:
```diff
@@ -0,0 +1,24 @@
+    <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.11.2" />
+    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.11.2" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="appsettings.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
```
```suggestion
```

evgenyfedorov2 replied:
Done and removed the file because it is not required for this app

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/trace-based/Program.cs:
```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
```suggestion
```

evgenyfedorov2 replied:
Done

IEvangelist commented on docs/core/extensions/snippets/logging/logging-sampling/trace-based/Program.cs:
```diff
@@ -0,0 +1,54 @@
+    private static readonly ActivitySource DemoSource = new("LogSamplingTraceBased");
+
+    public static void Main()
+    {
+        var hostBuilder = Host.CreateApplicationBuilder();
+        hostBuilder.Logging.AddSimpleConsole(options =>
+        {
+            options.SingleLine = true;
+            options.TimestampFormat = "hh:mm:ss";
+        });
+
+        // Add the Random probabilistic sampler to the logging pipeline.
+        hostBuilder.Logging.AddTraceBasedSampler();
+
+        var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            // enable Tracing sampling configured with 50% probability:
+            .SetSampler(new TraceIdRatioBasedSampler(0.5))
+            .AddSource("LogSamplingTraceBased")
+            .AddConsoleExporter()
+            .Build();
+
+        var app = hostBuilder.Build();
+        var loggerFactory = app.Services.GetRequiredService<ILoggerFactory>();
+        var logger = loggerFactory.CreateLogger("SamplingDemo");
+
+        // on average, 50% of Activities and logs will be sampled:
+        for (int i = 0; i < 10; i++)
+        {
+            using var activity = DemoSource.StartActivity("SayHello");
+            activity?.SetTag("foo", "bar");
+            activity?.SetStatus(ActivityStatusCode.Ok);
+
+            // the parent activity is sampled with 50% probability,
+            // and the same sampling decision will be used for logging
+            logger.NoisyMessage(i);
+        }
+
+        tracerProvider.Dispose();
+    }
+}
```
```suggestion
using ActivitySource demoSource = new("LogSamplingTraceBased");

var hostBuilder = Host.CreateApplicationBuilder(args);

hostBuilder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
hostBuilder.Logging.AddTraceBasedSampler();

using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    // Enable Tracing sampling configured with 50% probability:
    .SetSampler(new TraceIdRatioBasedSampler(0.5))
    .AddSource("LogSamplingTraceBased")
    .AddConsoleExporter()
    .Build();

using var app = hostBuilder.Build();

var loggerFactory = app.Services.GetRequiredService<ILoggerFactory>();
var logger = loggerFactory.CreateLogger("SamplingDemo");

// On average, 50% of Activities and logs will be sampled:
for (int i = 0; i < 10; i++)
{
    using var activity = demoSource.StartActivity("SayHello");
    activity?.SetTag("foo", "bar");
    activity?.SetStatus(ActivityStatusCode.Ok);

    // The parent activity is sampled with 50% probability,
    // and the same sampling decision will be used for logging
    logger.NoisyMessage(i);
}
```

evgenyfedorov2 replied:
Done

IEvangelist commented on docs/navigate/tools-diagnostics/toc.yml:
```diff
@@ -363,6 +363,8 @@ items:
         href: ../../core/diagnostics/logging-tracing.md
       - name: ILogger Logging
         href: ../../core/extensions/logging.md
+      - name: Logging Sampling
```
```suggestion
      - name: Log Sampling
```

evgenyfedorov2 replied:
Done

IEvangelist reviewed (commented):
Hi @evgenyfedorov2,

Let's get this cleaned up then have another quick look. This is awesome content, really looking forward to seeing this in docs. I'm curious, have you considered adding this to the [fundamentals TOC.yml](https://github.com/dotnet/docs/blob/main/docs/fundamentals/toc.yml#L1079) as well, so that it could be positioned under the Logging section there?

![image](https://github.com/user-attachments/assets/6f445b81-f54e-46a4-ab7a-1c49781a0367)

evgenyfedorov2 committed:
PR Comments

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/Program.cs:
```diff
@@ -0,0 +1,40 @@
+            Log.NoisyLogMessage(logger);
+            Log.LeftWhileLoop(logger);
+
+            Thread.Sleep(100);
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+---
+title: Logging sampling
+description: Learn how to fine-tune the volume of logs emitted by your application using Logging Sampling.
+ms.author: efedorov
```
Removed

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+---
+title: Logging sampling
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+---
+title: Logging sampling
+description: Learn how to fine-tune the volume of logs emitted by your application using Logging Sampling.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+ms.date: 04/29/2025
+---
+
+# Logging sampling in .NET
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+- Trace-based sampling - Sample logs based on the sampling decision of the current trace.
+- Random probabilistic sampling - Sample logs based on configured probability rules.
+- Custom sampling - Implement your own custom sampling strategy. See the [Implement custom sampling](#implement-custom-sampling) section for more details.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+- Random probabilistic sampling - Sample logs based on configured probability rules.
+- Custom sampling - Implement your own custom sampling strategy. See the [Implement custom sampling](#implement-custom-sampling) section for more details.
+
+However, please note that only one sampler can be used at a time. If you register multiple samplers, only the last one will be used.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+Logging sampling extends [filtering capabilities](logging.md#configure-logging-with-code), it allows for more fine-grained control over the logs emitted by your application -
+in addition to enabling or completely disabling logs, you can configure it to emit only a certain share of logs.
+In other words, from the probabilities 0 (= no logs) and 1 (= all logs) provided by the [filtering](logging.md#configure-logging-with-code),
+we can specify any value in between, e.g. 0.1 (= 10% of logs) or 0.25 (= 25% of logs).
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+## Configure trace-based sampling
+
+Trace-based sampling ensures that logs are sampled consistently with the underlying <xref:System.Diagnostics.Activity>. This is useful when you want to maintain correlation between traces and logs. You can enable tracing sampling as per the [guide](../diagnostics//distributed-tracing-concepts.md#sampling) and then enable the trace-based sampling:
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+There are several ways to configure random probabilistic sampling with its rules:
+
+### Using file configuration
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+### Using file configuration
+
+Create a configuration section in your `appsettings.json`, for example:
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+The configuration above:
+
+- Samples 25% of logs from categories starting with `Microsoft.AspNetCore.` of the <xref:Microsoft.Extensions.Logging.LogLevel.Information> level.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+> The <xref:Microsoft.Extensions.Diagnostics.Sampling.RandomProbabilisticSamplerFilterRule.Probability> value represents probability with values from 0 to 1.
+> For example, 0.25 means 25% of logs will be sampled. 0 means no logs will be sampled, and 1 means all logs will be sampled.
+> Those cases with 0 and 1 can be used to effectively disable or enable all logs for a specific rule.
+> Probability cannot be less than 0 or greater than 1, and if this occurs in the application, an exception will be thrown.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+:::code language="json" source="snippets/logging/logging-sampling/appsettings.noop.json" :::
+
+Then, while the application is running, you can update the `appsettings.json` to this:
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+:::code language="json" source="snippets/logging/logging-sampling/appsettings.updated.json" :::
+
+and the new rules will be applied automatically, for instance, with the configuration above, 1% of logs of the <xref:Microsoft.Extensions.Logging.LogLevel.Information> level will be sampled.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+- If multiple rules are selected, take the **last** one.
+- If no rules are selected, sampling is not applied, e.g. the log record is emitted as usual.
+
+### Using code configuration
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+:::code language="csharp" source="snippets/logging/logging-sampling/code-config/Program.cs" range="23-29":::
+
+The configuration above:
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+For example, you can create a custom sampler which can do, for instance:
+
+- Make a sampling decision based on whether a certain key/value pair is present in the log state or not, and what values it has.
+- Implement a rate-limiting strategy, e.g. only emit logs if the number of logs emitted in the pre-defined time interval is less than a certain threshold.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+To implement a custom sampler, follow these steps:
+
+1. Create a class that inherits from <xref:Microsoft.Extensions.Logging.LoggingSampler>.
+2. Implement the <<xref:Microsoft.Extensions.Logging.LoggingSampler.ShouldSample%2A?displayProperty=nameWithType> method to define your custom sampling logic.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+1. Create a class that inherits from <xref:Microsoft.Extensions.Logging.LoggingSampler>.
+2. Implement the <<xref:Microsoft.Extensions.Logging.LoggingSampler.ShouldSample%2A?displayProperty=nameWithType> method to define your custom sampling logic.
+3. Register your custom sampler in the logging pipeline using the <xref:Microsoft.Extensions.Logging.SamplingLoggerBuilderExtensions.AddSampler%2A> extension method.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+
+For each log record (if it is not filtered out), the <xref:Microsoft.Extensions.Logging.LoggingSampler.ShouldSample%2A?displayProperty=nameWithType>
+method will be called exactly once and its return value will be used to determine if the log
+records should be emitted or not.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+## Performance considerations
+
+Logging sampling is optimized for storage at the cost of slightly increased CPU usage. In case your application produces a lot of logs whose storage costs you a lot of money,
+you can use sampling to reduce the volume of logs. If configured correctly, you will not lose information important for incident resolution, but you will reduce the volume of logs and hence the cost of storage.
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+| <xref:Microsoft.Extensions.Logging.LogLevel.Information> | Do apply sampling |
+| <xref:Microsoft.Extensions.Logging.LogLevel.Warning>   | Consider applying sampling |
+| <xref:Microsoft.Extensions.Logging.LogLevel.Error>     | Do not apply sampling |
+| <xref:Microsoft.Extensions.Logging.LogLevel.Critical>  | Do not apply sampling |
```
Done

evgenyfedorov2 commented on docs/core/extensions/logging-sampling.md:
```diff
@@ -0,0 +1,183 @@
+- Use category-based rules to target specific components.
+- Consider using trace-based sampling if you're using distributed tracing.
+- Monitor the effectiveness of your sampling rules together.
+- Find the balance specific to your application - very low sampling rates can degrade observability, while high sampling rates can increase costs.
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/Program.cs:
```diff
@@ -0,0 +1,40 @@
+            Thread.Sleep(100);
+        }
+    }
+}
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/appsettings.json:
```diff
@@ -0,0 +1,25 @@
+      "Default": "Debug"
+    }
+  },
+
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/appsettings.noop.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "Logging": {
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/code-config/Log.cs:
```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/code-config/LogSamplingCodeConfig.csproj:
```diff
@@ -0,0 +1,21 @@
+    <None Update="appsettings.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/code-config/Program.cs:
```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/code-config/Program.cs:
```diff
@@ -0,0 +1,40 @@
+            Log.NoisyMessage(logger);
+        }
+    }
+}
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/file-config/Log.cs:
```diff
@@ -0,0 +1,18 @@
+internal static partial class Log
+{
+    [LoggerMessage(Level = LogLevel.Debug, Message = "Entered While loop in my application.")]
+    public static partial void EnteredWhileLoop(ILogger logger);
```
Done

evgenyfedorov2 commented on docs/core/extensions/snippets/logging/logging-sampling/file-config/LogSamplingFileConfig.csproj:
```diff
@@ -0,0 +1,21 @@
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Hosting" Version="9.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="9.0.0" />
```
Done

evgenyfedorov2 committed:
Fix links

evgenyfedorov2 commented:
@IEvangelist Thank you for the thorough review! I have addressed all comments, please take a look.

IEvangelist commented on docs/core/extensions/snippets/logging/log-sampling/Program.cs:
```diff
@@ -0,0 +1,29 @@
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+var hostBuilder = Host.CreateApplicationBuilder();
+hostBuilder.Logging.AddSimpleConsole(options =>
+{
+    options.SingleLine = true;
+    options.TimestampFormat = "hh:mm:ss";
+});
+
+// Add the Random probabilistic sampler to the logging pipeline.
+hostBuilder.Logging.AddRandomProbabilisticSampler(0.01, LogLevel.Information);
+hostBuilder.Logging.AddRandomProbabilisticSampler(0.1, LogLevel.Warning);
```
```suggestion
var builder = Host.CreateApplicationBuilder();
builder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
builder.Logging.AddRandomProbabilisticSampler(0.01, LogLevel.Information);
builder.Logging.AddRandomProbabilisticSampler(0.1, LogLevel.Warning);
```

IEvangelist commented on docs/core/extensions/snippets/logging/log-sampling/Program.cs:
```diff
@@ -0,0 +1,29 @@
+hostBuilder.Logging.AddRandomProbabilisticSampler(0.01, LogLevel.Information);
+hostBuilder.Logging.AddRandomProbabilisticSampler(0.1, LogLevel.Warning);
+
+using var app = hostBuilder.Build();
```
```suggestion
using var app = builder.Build();
```

IEvangelist commented on docs/core/extensions/snippets/logging/log-sampling/code-config/Program.cs:
```diff
@@ -0,0 +1,32 @@
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+var hostBuilder = Host.CreateApplicationBuilder(args);
+
+hostBuilder.Logging.AddSimpleConsole(options =>
+{
+    options.SingleLine = true;
+    options.TimestampFormat = "hh:mm:ss";
+});
+
+// Add the Random probabilistic sampler to the logging pipeline.
+hostBuilder.Logging.AddRandomProbabilisticSampler(options =>
+{
+    options.Rules.Add(
+        new RandomProbabilisticSamplerFilterRule(
+            probability: 0.05d,
+            eventId : 1001));
+});
+
+using var app = hostBuilder.Build();
```
```suggestion
var builder = Host.CreateApplicationBuilder(args);

builder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
builder.Logging.AddRandomProbabilisticSampler(options =>
{
    options.Rules.Add(
        new RandomProbabilisticSamplerFilterRule(
            probability: 0.05d,
            eventId : 1001));
});

using var app = builder.Build();
```

IEvangelist commented on docs/core/extensions/snippets/logging/log-sampling/file-config/Program.cs:
```diff
@@ -0,0 +1,29 @@
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+var hostBuilder = Host.CreateApplicationBuilder();
+
+hostBuilder.Logging.AddSimpleConsole(options =>
+{
+    options.SingleLine = true;
+    options.TimestampFormat = "hh:mm:ss";
+});
+
+// Add the Random probabilistic sampler to the logging pipeline.
+hostBuilder.Logging.AddRandomProbabilisticSampler(hostBuilder.Configuration);
+
+using var app = hostBuilder.Build();
```
```suggestion
var builder = Host.CreateApplicationBuilder();

builder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
builder.Logging.AddRandomProbabilisticSampler(builder.Configuration);

using var app = builder.Build();
```

IEvangelist commented on docs/core/extensions/snippets/logging/log-sampling/trace-based/Program.cs:
```diff
@@ -0,0 +1,44 @@
+
+using ActivitySource demoSource = new("LogSamplingTraceBased");
+
+var hostBuilder = Host.CreateApplicationBuilder(args);
+
+hostBuilder.Logging.AddSimpleConsole(options =>
+{
+    options.SingleLine = true;
+    options.TimestampFormat = "hh:mm:ss";
+});
+
+// Add the Random probabilistic sampler to the logging pipeline.
+hostBuilder.Logging.AddTraceBasedSampler();
+
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+    // Enable Tracing sampling configured with 50% probability:
+    .SetSampler(new TraceIdRatioBasedSampler(0.5))
+    .AddSource("LogSamplingTraceBased")
+    .AddConsoleExporter()
+    .Build();
+
+using var app = hostBuilder.Build();
```
```suggestion
var builder = Host.CreateApplicationBuilder(args);

builder.Logging.AddSimpleConsole(options =>
{
    options.SingleLine = true;
    options.TimestampFormat = "hh:mm:ss";
});

// Add the Random probabilistic sampler to the logging pipeline.
builder.Logging.AddTraceBasedSampler();

using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    // Enable Tracing sampling configured with 50% probability:
    .SetSampler(new TraceIdRatioBasedSampler(0.5))
    .AddSource("LogSamplingTraceBased")
    .AddConsoleExporter()
    .Build();

using var app = builder.Build();
```

IEvangelist committed:
Apply suggestions from code review

Minor simplification of host builder variable name, to be consistent with fundamentals docs.
